### PR TITLE
Update py27/py34 deprecation docs

### LINKED ISSUE #5275: Technical aspects related to Python 2.7 and 3.4 support (pytest 5.0)
This issue is to discuss the technical aspects of pytest dropping Python 2 and 3.4 in pytest 5.0 and long term support of the 4.6 series.

For a more user-centric point of view, please see [Python 2.7 and 3.4 support plan](https://docs.pytest.org/en/latest/py27-py34-deprecation.html) in the docs.

---

(This post has since then edited from the original version after the discussion that took place here).

Unless decided otherwise, the last pytest version to support **Python 2.7 and 3.4** will be the **4.5** series, which is the last minor release on PyPI. 

This issue contains the agreed process on how we will proceed with the pytest 5.0 release and long term support plans for the 4.5 series. 

## What goes into 4.6.X releases

New 4.6.X releases will contain *bug fixes* only:

* A bug *will* be fixed/ported if, in a normal release, we would merge it in the `master` branch: often we apply bug fixes to the `features` branch because even though we are fixing a bug, it may break test suites, so we often take the safest route. Those bugs are often not critical, and won't be ported to the maintenance branch.
* No new features will be added on this branch, however small. 

## Who will handle porting bugs

After a PR for a bug fix is merged, and the bug is applicable to the 4.6.X series according to the rules stated previously:

We expect **core maintainers** who provided the initial bug fix to also create a new PR porting the bug.

From **other contributors** we expect core maintainers to step up to create a new PR porting the bug if the original contributor is not interested in doing so. @nicoddemus and @asottile applied to be the forefront here, but will gladly accept help from other core maintainers.

## When will 4.6.X releases happen

New 4.6.X releases will happen after we have a few bugs in place to release, or if a few weeks have passed (say a single bug has been fixed a month after the latest 4.5.X release). No hard rules here (as with any other pytest release really), just a ballpark.

## Backporting changes into the 4.6 release

1. `git fetch --all --prune`
2. `git checkout origin/4.6-maintenance -b backport-XXXX`  # use the PR number here
3. Locate the merge commit on the PR, in the "merged" message, for example:

   > nicoddemus merged commit 0f8b462 into pytest-dev:features

4. `git cherry-pick -m1 REVISION`  # use the revision you found in 3.
5. open a PR targeting `4.6-maintenance`: prefix the message with `[4.6]` so it is an obvious backport, delete the PR body (it usually contains a duplicated commit message)

Note for backport PRs a review is not necessary -- though if it it falls into one of these categories a review is strongly suggested:
- The cherry-pick did not apply cleanly
- This is a bugfix for only the 4.6.x series

# Steps for the first 5.0 release

1. The absolute first step is to mark `features` as Python 3 only: ✔️  
    * Update `setup.py` classifiers and `python_requires` to declare pytest only supports Python 3.5+.
    * Drop Python<3.5 from CI.
    * Update the "Python 2.7 and 3.4 support plan".  
1. After `4.6` is released, we push a `pytest4-maintenance` branch pointing to `4.6.0` initially. ✔️
2. The next step is to run `pyupgrade --py3-plus` (I guess @asottile plans to volunteer here :wink:); ✔️
1. Turn warnings into errors. ✔️
1. Release 5.0. 🎉

## When will we start the 5.0 release process

It is my opinion that we can proceed with the steps above right away, as we are quickly approaching end of May.

If nobody opposes, I plan to start with the first step this week or early next week.

### PR DESCRIPTION
**Pull request by nicoddemus**
I've updated the text and incorporated the topics from #5275, so this
can now be part of the official docs, and #5275 can be closed/unpinned.

Closes #5275

### CODE REVIEW

File: doc/en/py27-py34-deprecation.rst [hunk 1]
```diff
@@ -1,31 +1,96 @@
...
 Thanks to the `python_requires`_ ``setuptools`` option,
 Python 2.7 and Python 3.4 users using a modern ``pip`` version
-will install the last pytest ``4.6`` version automatically even if ``5.0`` or later
+will install the last pytest ``4.6`` version automatically even if ``5.0`` or later versions
 are available on PyPI.
 
-While pytest ``5.0`` will be the new mainstream and development version, until **January 2020**
-the pytest core team plans to make bug-fix releases of the pytest ``4.6`` series by
-back-porting patches to the ``4.6-maintenance`` branch that affect Python 2 users.
+Users should ensure they are using the latest ``pip`` and ``setuptool`` versions for this to work.
```
Comment (asottile) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
Users should ensure they are using the latest ``pip`` and ``setuptools`` versions for this to work.
```

File: doc/en/py27-py34-deprecation.rst [hunk 2]
```diff
@@ -1,31 +1,96 @@
...
+
+For those reasons, in Jun 2019, it was decided that **pytest 4.6** series will be the last to support Python 2.7 and 3.4.
+
+What this means for general users
+---------------------------------
 
 Thanks to the `python_requires`_ ``setuptools`` option,
 Python 2.7 and Python 3.4 users using a modern ``pip`` version
-will install the last pytest ``4.6`` version automatically even if ``5.0`` or later
+will install the last pytest ``4.6`` version automatically even if ``5.0`` or later versions
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]: No need for code formatting for version numbers, it's not done for "Python 2.7" or "**pytest 4.6**" or "4.6.X" in some places. 

```suggestion
will install the last pytest 4.6 version automatically even if 5.0 or later versions
```

File: doc/en/py27-py34-deprecation.rst (hunk 1 above)
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]: No need for code formatting for package/project names, it's not done for "pytest".
 
```suggestion
Users should ensure they are using the latest pip and setuptools versions for this to work.
```

File: doc/en/py27-py34-deprecation.rst [hunk 3]
```diff
@@ -1,31 +1,96 @@
...
 
-While pytest ``5.0`` will be the new mainstream and development version, until **January 2020**
-the pytest core team plans to make bug-fix releases of the pytest ``4.6`` series by
-back-porting patches to the ``4.6-maintenance`` branch that affect Python 2 users.
+Users should ensure they are using the latest ``pip`` and ``setuptool`` versions for this to work.
+
+Maintenance of 4.6.X versions
+-----------------------------
+
+Until Jan 2020, the pytest core team ported many bug-fixes from the main release into the
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
Until January 2020, the pytest core team ported many bug-fixes from the main release into the
```

File: doc/en/py27-py34-deprecation.rst [hunk 4]
```diff
@@ -1,31 +1,96 @@
...
+``4.6-maintenance`` branch, with several ``4.6.X`` releases being made along the year.
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
``4.6-maintenance`` branch, with several 4.6.X releases being made along the year.
```

File: doc/en/py27-py34-deprecation.rst [hunk 5]
```diff
@@ -1,31 +1,96 @@
...
+
+From now on, the core team will **no longer actively backport patches**, but the ``4.6-maintenance``
+branch will continue to exist so the community itself can contribute patches.
 
-**After 2020**, the core team will no longer actively backport patches, but the ``4.6-maintenance``
-branch will continue to exist so the community itself can contribute patches. The core team will
-be happy to accept those patches and make new ``4.6`` releases **until mid-2020**.
+The core team will be happy to accept those patches and make new ``4.6.X`` releases **until mid-2020**;
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
The core team will be happy to accept those patches and make new 4.6.X releases **until mid-2020**;
```

File: doc/en/py27-py34-deprecation.rst [hunk 6]
```diff
@@ -1,31 +1,96 @@
...
+passed (say a single bug has been fixed a month after the latest 4.6.X release).
+No hard rules here, just a ballpark.
+
+Who will handle applying bug fixes
+++++++++++++++++++++++++++++++++++
+
+We core maintainers expect that people still using Python 2.7/3.4 and being affected by
+bugs to step up and provide patches and/or port bug fixes from the active branches.
+
+We will be happy to guide users interested in doing so, so please don't hesitate on asking.
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
We will be happy to guide users interested in doing so, so please don't hesitate to ask.
```

File: doc/en/py27-py34-deprecation.rst [hunk 7]
```diff
@@ -1,31 +1,96 @@
...
+#. ``git cherry-pick -m1 REVISION`` # use the revision you found above (``0f8b462``).
+
+#. Open a PR targeting ``4.6-maintenance``:
+
+   * Prefix the message with ``[4.6]`` so it is an obvious backport
+   * Delete the PR body, it usually contains a duplicated commit message.
+
+**Providing new PRs to 4.6**
+
+Fresh pull requests to the ``4.6-maintenance`` will be accepted provided that
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
Fresh pull requests to ``4.6-maintenance`` will be accepted provided that
```

File: doc/en/py27-py34-deprecation.rst [hunk 8]
```diff
@@ -1,31 +1,96 @@
...
+the similar code in the active branches does not contain that bug (for example, is a bug specific
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
the equivalent code in the active branches does not contain that bug (for example, a bug is specific
```

File: doc/en/py27-py34-deprecation.rst [hunk 9]
```diff
@@ -1,31 +1,96 @@
...
+to Python 2 only).
 
-The technical aspects of the Python 2.7 and 3.4 support plan (such as when releases will occur, how to backport fixes, etc) is described in issue `#5275 <https://github.com/pytest-dev/pytest/issues/5275>`__.
+Bug fixes that also happen in the main stream version should be first fixed
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
Bug fixes that also happen in the mainstream version should be first fixed
```

File: doc/en/py27-py34-deprecation.rst [hunk 10]
```diff
@@ -1,31 +1,96 @@
...
+there, and then backported as per-instructions above.
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
there, and then backported as per instructions above.
```

File: doc/en/py27-py34-deprecation.rst [hunk 11]
```diff
@@ -1,31 +1,96 @@
...
+#. Locate the merge commit on the PR, in the "merged" message, for example:
+
+    nicoddemus merged commit 0f8b462 into pytest-dev:features
+
+#. ``git cherry-pick -m1 REVISION`` # use the revision you found above (``0f8b462``).
+
+#. Open a PR targeting ``4.6-maintenance``:
+
+   * Prefix the message with ``[4.6]`` so it is an obvious backport
+   * Delete the PR body, it usually contains a duplicated commit message.
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
   * Delete the PR body, it usually contains a duplicate commit message
```

File: doc/en/py27-py34-deprecation.rst [hunk 12]
```diff
@@ -1,31 +1,96 @@
...
++++++++++++++++++++++++++++++
+
+New 4.6.X releases will contain bug fixes only.
+
+When will 4.6.X releases happen
++++++++++++++++++++++++++++++++
+
+New 4.6.X releases will happen after we have a few bugs in place to release, or if a few weeks have
+passed (say a single bug has been fixed a month after the latest 4.6.X release).
+No hard rules here, just a ballpark.
```
Comment (hugovk) [inline doc/en/py27-py34-deprecation.rst]:
```suggestion
No hard rules here, just ballpark.
```

**Comment by nicoddemus** [comment]
Thanks @hugovk, appreciate it!

**Pull request merged**